tests/framework/extensions/namespaces/namespaces.go

```diff
@@ -0,0 +1,100 @@
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	"github.com/rancher/rancher/tests/framework/pkg/convert"
+	"github.com/rancher/rancher/tests/framework/pkg/wait"
+	"github.com/rancher/rancher/tests/integration/pkg/defaults"
+	coreV1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+var NamespaceGroupVersionResource = schema.GroupVersionResource{
+	Group:    "",
+	Version:  "v1",
+	Resource: "namespaces",
+}
+
+// ContainerDefaultResourceLimit sets the container default resource limit in a string
+// limitsCPU and requestsCPU in form of "3m"
+// limitsMemory and requestsMemory in the form of "3Mi"
+func ContainerDefaultResourceLimit(limitsCPU, limitsMemory, requestsCPU, requestsMemory string) string {
+	containerDefaultResourceLimit := fmt.Sprintf("{\"limitsCpu\": \"%s\", \"limitsMemory\":\"%s\",\"requestsCpu\":\"%s\",\"requestsMemory\":\"%s\"}",
+		limitsCPU, limitsMemory, requestsCPU, requestsMemory)
+	return containerDefaultResourceLimit
+}
+
+// NewProjectNamespace is a constructor that sets up a namespace object to be used by the GetDownStreamK8Client to create a project namespace
+func NewProjectNamespace(namespaceName, projectID, containerDefaultResourceLimit string, labels, annotations map[string]string) *coreV1.Namespace {
+	annotations["field.cattle.io/containerDefaultResourceLimit"] = containerDefaultResourceLimit
+	annotations["field.cattle.io/projectId"] = projectID
+	namespace := &coreV1.Namespace{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Namespace",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        namespaceName,
+			Annotations: annotations,
+			Labels:      labels,
+		},
+	}
+	return namespace
+}
+
+func CreateNamespace(client *rancher.Client, namespace *coreV1.Namespace) (*coreV1.Namespace, error) {
+	dynamicClient, err := client.GetRancherDynamicClient()
+	if err != nil {
+		return nil, err
+	}
+
+	namespaceResource := dynamicClient.Resource(NamespaceGroupVersionResource).Namespace("")
+
+	unstructuredObject, err := convert.ToUnstructured(namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	unstructuredResp, err := namespaceResource.Create(context.TODO(), unstructuredObject, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	client.Session.RegisterCleanupFunc(func() error {
+		err := namespaceResource.Delete(context.TODO(), unstructuredResp.GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+
+		watchInterface, err := namespaceResource.Watch(context.TODO(), metav1.ListOptions{
+			FieldSelector:  "metadata.name=" + unstructuredResp.GetName(),
+			TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		return wait.WatchWait(watchInterface, func(event watch.Event) (ready bool, err error) {
+			if event.Type == watch.Error {
+				return false, fmt.Errorf("there was an error deleting cluster")
```

Comment:
isn't the error handled by the `WatchWait` function?

Reply:
Not currently, but I can move the error check there instead if that makes more sense.

Reply:
Actually I think that something our check function should check that way we can have a more detailed error message. If we error check in `WatchWait` we won't know the specific error.